rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+use crate::ffi;
+use std::borrow::Borrow;
+use std::ffi::c_char;
+use std::ffi::c_int;
+use std::ffi::c_uint;
+use std::ffi::CStr;
+use std::ffi::OsStr;
+#[cfg(windows)]
+use std::mem::forget;
+use std::mem::ManuallyDrop;
+use std::ptr::null;
+use std::ptr::NonNull;
+use std::str::Utf8Error;
+
+// This is more or less equivalent to manually defining Display and From<other error types>
+use thiserror::Error;
+
+/// A loaded shared library for a stan model
+pub struct StanLibrary(ManuallyDrop<ffi::Bridgestan>);
+
+#[cfg(windows)]
+impl Drop for StanLibrary {
+    fn drop(&mut self) {
+        let lib = unsafe { ManuallyDrop::take(&mut self.0) };
+        forget(lib.into_library());
+    }
+}
+
+#[cfg(not(windows))]
+impl Drop for StanLibrary {
+    fn drop(&mut self) {
+        unsafe { ManuallyDrop::drop(&mut self.0) };
+    }
+}
+
+/// A callback for print statements in stan models
+pub type StanPrintCallback = extern "C" fn(*const c_char, usize);
+
+impl StanLibrary {
+    /// Provide a callback function to be called when stan prints a message
+    ///
+    /// The provided function must never panic.
+    ///
+    /// Since the call is proteted by a mutex internally, it does not
+    /// need to be thread safe.
+    pub unsafe fn set_print_callback(&mut self, callback: StanPrintCallback) -> Result<()> {
+        let mut err = ErrorMsg::new(&self);
+        let rc = unsafe { self.0.bs_set_print_callback(Some(callback), err.as_ptr()) };
+
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+}
+
+/// Error type for bridgestan interface
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum BridgeStanError {
```

Comment:
What is our API contract here? I've read various things about the dangers of re-exporting error types from libraries because a major version bump in those libraries means you also need one - does that apply to `thiserror` and the `#[from]` macro, or does it avoid that issue?

Reply:
Good point. I changed it to `InvalidLibrary(String)`.

Reply:
Does the `From` impl not create the same issue? The exact boundary of what is a public API isn't fully clear to me.

Reply:
You are right. Since the `From` trait and `BridgeStanError` are both visible from the client crate, the impl `From<libloading::Error> for BridgeStanError` will also be visible. So a user could in theory decide to create their own libloading error and convert that to a bridgestan error. If we were then to change to some other dynamic loading library, that code would break.

We can pretty easily avoid that by creating a `LoadingError` type that contains a private libloading error.  (just pushed a commit).

Reply:
Sounds good (Surely `LoadingError` shouldn't be `pub`, though? At least going off the thiserror example of a similar idea)

Reply:
Unfortunately it has to be, because BridgeStanError is an enum, and all variants of a public enum have to be public. We don't export it however, so it is unnamable (so user code can never create one).
This whole visibility stuff is one of the parts of rust I dislike most. It really is confusing.

Reply:
Ah, the disconnect between `pub` and actually being exported was what had me hung up. Thanks!